### PR TITLE
[Feature]Add Mega moe support for a5

### PR DESCRIPTION
**Pull request by hanxi-java**
[Feature]Add Mega moe support for a5 (#14658)

### What this PR does / why we need it?

Enable MegaMoe's MXFP quantization path on Ascend 950 (A5). A5 MegaMoe is
MXFP-only, so this adds the config gating, communication-method selection,
quant settings, and weight-layout handling required to run W8A8MXFP /
W4A8MXFP / W4A4MXFP on A5, while keeping the existing A2/A3 INT path and
the unquantized BF16 path unchanged.

- Add A5-specific MegaMoe config validation (hidden / intermediate /
  expert-count / top-k constraints) and stop silently downgrading
  enable_fused_mc2 for unsupported configs.
- Thread is_draft_model through select_moe_comm_method so MTP draft models
  skip FUSED_MC2 on A5; extract is_mega_moe_enabled() and
  can_use_mega_moe_in_current_forward() helpers.
- Add MXFP dispatch_quant_mode and FP8/FP4 dtype enums, and map
  W8A8MXFP / W4A8MXFP / W4A4MXFP in _get_cann_mega_moe_quant_settings.
- Normalize MegaMoe weight/scale/bias inputs to per-expert lists and force
  x_active_mask to None on A5.
- Fix MegaMoe symm-buffer sizing (use scheduler max_num_tokens; fall back
  to logical/local expert counts when num_experts is unset).
- Route the unquantized BF16 MoE path to dispatch_ffn_combine instead of
  MegaMoe.

### Does this PR introduce _any_ user-facing change?

# Accuracy Test Report

## Summary Overview
- **Model**: `Qwen3.5-397B-A17B`
- **Dataset**: gpqa
- **Total Test Runs**: 3
- **Average Accuracy**: 85.61%
- **Evaluation Result**: Passed (Meets Standard)

---

## Detailed Test Results

| Run Number | Accuracy (%) |
| :---: | :---: |
| Run 1 | 86.36% |
| Run 2 | 85.62% |
| Run 3 | 84.85% |
| **Average** | **85.61%** |

No
### How was this patch tested?
https://github.com/vllm-project/vllm/commit/b2f685834a6456197e7033966fdef52a23f1abcd

Signed-off-by: hanxi-java <634498162@qq.com>

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/ops/fused_moe/moe_utils.py [hunk 1]
```diff
@@ -31,6 +31,14 @@
...
 _CANN_MEGA_MOE_QUANT_MODE_None = 0
 _CANN_MEGA_MOE_QUANT_MODE_INT8 = 2
+_CANN_MEGA_MOE_QUANT_MODE_MXFP = 4
+# FP8/FP4 dispatch-quant-out dtype and weight-type enums (ACL/CANN). The FP8
+# values match mega_moe.py's _dtype_to_int and the Ascend 950 MegaMoe
+# dispatchQuantOutDtype values documented in cann_ops_transformer
+# (23 = FLOAT8_E5M2, 24 = FLOAT8_E4M3FN, 296 = FLOAT4_E2M1).
+_CANN_ACL_FLOAT8_E5M2 = 23
+_CANN_ACL_FLOAT8_E4M3FN = 24
+_CANN_ACL_FLOAT4_E2M1 = 296
```
Comment (weiguihua2) [inline vllm_ascend/ops/fused_moe/moe_utils.py:44]: A description of the constants is needed.

File: vllm_ascend/quantization/methods/w4a8.py
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/w4a8.py:1]: A5 mx quant does not involve this file.

File: vllm_ascend/quantization/methods/w8a8_dynamic.py
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/w8a8_dynamic.py:1]: the same as w4a8.py, plz check

File: vllm_ascend/quantization/methods/w8a8_mxfp8.py [hunk 2]
```diff
@@ -349,6 +367,17 @@ def process_weights_after_loading(self, layer):
...
+
+        # MegaMoe (FUSED_MC2 on A5) expects the weights in their original
+        # (out, in) layout — weight1=(E, 2*inter, hidden), weight2=(E, hidden,
+        # inter) — and the reshaped (not transposed) E8M0 scales. Capture
+        # per-expert lists before the non-mega path transposes everything below.
+        if is_mega_moe_enabled():
+            layer.cann_mega_moe_w13_weight_list = list(layer.w13_weight.data.unbind(dim=0))
+            layer.cann_mega_moe_w2_weight_list = list(layer.w2_weight.data.unbind(dim=0))
+            layer.cann_mega_moe_w13_weight_scale_list = list(layer.w13_weight_scale.data.unbind(dim=0))
+            layer.cann_mega_moe_w2_weight_scale_list = list(layer.w2_weight_scale.data.unbind(dim=0))
```
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/w8a8_mxfp8.py]: When new weight tensors are created for megamoe, the original weight tensor should be deleted to save HBM memory.

File: vllm_ascend/ascend_forward_context.py [hunk 3]
```diff
@@ -277,7 +282,11 @@ def _select_a5_moe_comm_method(
     return MoECommType.ALLTOALL
 
 
-def select_moe_comm_method(num_tokens: int, vllm_config: VllmConfig) -> MoECommType | None:
+def select_moe_comm_method(
+    num_tokens: int,
+    vllm_config: VllmConfig,
+    is_draft_model: bool = False,
```
Comment (weijinqian0) [inline vllm_ascend/ascend_forward_context.py:401]: Do not add unnecessary parameters. The selected communication mode should not be related to the draft_model.

File: vllm_ascend/ops/fused_moe/routed_experts.py [hunk 4]
```diff
@@ -90,18 +90,16 @@ def process_weights_after_loading(self, layer):
         # in their native format without explicit casting here.
         enable_fused_mc2 = get_ascend_config().enable_fused_mc2
         if enable_fused_mc2:
-            if _MEGA_MOE_SUPPORTED:
+            # MegaMoe is MXFP-only (FP8/FP4), so the unquantized path always
+            # uses the dispatch_ffn_combine op with FRACTAL_NZ weights.
+            layer.w13_weight.data = torch_npu.npu_format_cast(layer.w13_weight.data, ACL_FORMAT_FRACTAL_NZ)
+            layer.w2_weight.data = torch_npu.npu_format_cast(layer.w2_weight.data, ACL_FORMAT_FRACTAL_NZ)
+            if enable_fused_mc2 == 1 and self.dynamic_eplb:
```
Comment (weijinqian0) [inline vllm_ascend/ops/fused_moe/routed_experts.py]: plz check this.

File: vllm_ascend/ops/fused_moe/moe_mlp.py [hunk 5]
```diff
@@ -594,7 +594,7 @@ def unquant_apply_mlp(
 
     gate_up_out = torch_npu.npu_grouped_matmul(
         x=[hidden_states],
-        weight=[w1],
+        weight=w1 if isinstance(w1, list) else [w1],
```
Comment (lcfenglinwan) [inline vllm_ascend/ops/fused_moe/moe_mlp.py]: Why should this be modified?

File: vllm_ascend/ops/fused_moe/moe_mlp.py [hunk 6]
```diff
@@ -667,7 +667,7 @@ def unquant_apply_mlp(
 
     hidden_states = torch_npu.npu_grouped_matmul(
         x=[gate_up_out],
-        weight=[w2],
+        weight=w2 if isinstance(w2, list) else [w2],
```
Comment (lcfenglinwan) [inline vllm_ascend/ops/fused_moe/moe_mlp.py]: same as above

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py [hunk 7]
```diff
@@ -185,19 +185,85 @@ def apply(
...
+                "w13_weight_scale",
+                "w2_weight_scale",
+            )
+            for tensor_name in tensor_names:
+                delattr(layer, tensor_name)
+        else:
+            # Non-mega path (npu_grouped_matmul): format-cast to FRACTAL_NZ and
+            # transpose to (in, out) so the packed FP4 K is interpreted correctly.
+            layer.w13_weight.data = torch_npu.npu_format_cast(
+                layer.w13_weight.data, 29, customize_dtype=torch.float8_e4m3fn, input_dtype=torch_npu.float4_e2m1fn_x2
```
Comment (lijiahang226) [inline vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py:258]: Replace 29 with ACL_FORMAT_FRACTAL_NZ.

**Comment by lijiahang226** [comment]
Please review this PR. @weijinqian0

File: vllm_ascend/quantization/methods/w8a8/w8a8_mxfp8.py [hunk 8]
```diff
@@ -365,11 +388,36 @@ def process_weights_after_loading(self, layer):
...
+            ]
+
+            tensor_names = (
+                "w13_weight",
+                "w2_weight",
+                "w13_weight_scale",
+                "w2_weight_scale",
+            )
+            for tensor_name in tensor_names:
+                delattr(layer, tensor_name)
```
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/w8a8/w8a8_mxfp8.py:414]: It is recommended to use the `dispose_tensor` function defined in `utils.py`.

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py [hunk 9]
```diff
@@ -185,19 +185,85 @@ def apply(
...
+        layer.w13_weight_scale.data = layer.w13_weight_scale.data.reshape(g, n, k // 2, 2)
         g, n, k = layer.w2_weight_scale.shape
-        layer.w2_weight_scale.data = layer.w2_weight_scale.data.reshape(g, n, k // 2, 2).transpose(-3, -2)
+        layer.w2_weight_scale.data = layer.w2_weight_scale.data.reshape(g, n, k // 2, 2)
+        # MegaMoe (FUSED_MC2 on A5) expects the weights in their original
+        # (out, in) layout — weight1=(E, 2*inter, hidden//2), weight2=(E,
+        # hidden, inter//2) for packed FP4 — and the reshaped (not transposed)
+        # scales. Capture per-expert FRACTAL_NZ lists before the non-mega path
+        # below format-casts and transposes everything.
+        if use_cann_megamoe(get_current_vllm_config()) and is_950():
```
Comment (lcfenglinwan) [inline vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py:221]: The `is_950()` here can probably be removed.

File: vllm_ascend/ascend_forward_context.py [hunk 10]
```diff
@@ -97,6 +98,40 @@ def use_cann_megamoe(vllm_config: VllmConfig) -> bool:
...
+def set_draft_moe_quant_type(draft_model: torch.nn.Module | None) -> None:
+    """Cache the draft model's MoE quantization type once, at process level.
+
+    Whether the MTP draft MoE layers can use the A5 mega moe (FUSED_MC2)
+    operator depends on their actual quantization, which is fixed for the
+    lifetime of the process once the draft model is loaded. Scan the draft
+    model's routed experts once and cache the result so ``select_moe_comm_method``
+    can consult it without touching the layer graph.
+    """
+    global _DRAFT_MOE_QUANT_TYPE
```
Comment (lijiahang226) [inline vllm_ascend/ascend_forward_context.py:116]: Introducing a global variable is generally discouraged. Could you please reconsider this part of the change?

File: vllm_ascend/quantization/methods/w4a8.py
Comment (hanxi-java) [reply vllm_ascend/quantization/methods/w4a8.py:1]: A8W4 is supported on A5.  Please Read doc on https://gitcode.com/cann/ops-transformer/blob/master/mc2/mega_moe/docs/torchapi_mega_moe.md.

File: vllm_ascend/quantization/methods/w8a8_dynamic.py
Comment (hanxi-java) [reply vllm_ascend/quantization/methods/w8a8_dynamic.py:1]: A8W8 is supported on A5. Please Read doc on https://gitcode.com/cann/ops-transformer/blob/master/mc2/mega_moe/docs/torchapi_mega_moe.md.

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py (hunk 9 above)
Comment (hanxi-java) [reply vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py]: W4W8-FP is only supported on A5 device, and only W4W8-INT is supported on A2/A3.Noted on https://gitcode.com/cann/ops-transformer/blob/master/mc2/mega_moe/docs/torchapi_mega_moe.md.

**Comment by hanxi-java** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by weijinqian0 — DISMISSED** [review]

**Comment by hanxi-java** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

File: vllm_ascend/ops/fused_moe/moe_comm_method.py [hunk 11]
```diff
@@ -416,17 +420,22 @@ def to_list(x):
...
-            # not already match — saves the kernel launch when an upstream
-            # change ever flips the reserved buffer to int8.
-            raw_mask = fused_experts_input.routing.mc2_mask
-            if raw_mask.dtype == torch.int8:
-                x_active_mask = raw_mask.contiguous()
-            else:
-                x_active_mask = raw_mask.to(torch.int8).contiguous()
+        # Ascend 950 (A5) MegaMoe only support a null x_active_mask, and it
+        # must be passed as None. But on A2/A3 it must be valid.
+        if is_950():
```
Comment (kunpengW-code) [inline vllm_ascend/ops/fused_moe/moe_comm_method.py]: Avoid using `is_950`. Please implement this via `hardware_profile`.

File: vllm_ascend/ops/fused_moe/routed_experts.py [hunk 12]
```diff
@@ -118,11 +118,20 @@ def process_weights_after_loading(self, layer):
                 layer.w13_weight.data = torch_npu.npu_format_cast(layer.w13_weight.data, ACL_FORMAT_FRACTAL_NZ)
                 layer.w2_weight.data = torch_npu.npu_format_cast(layer.w2_weight.data, ACL_FORMAT_FRACTAL_NZ)
             if use_megamoe or self.dynamic_eplb:
-                layer.w13_weight_list = [weight.clone() for weight in layer.w13_weight.data.unbind(dim=0)]
-                layer.w2_weight_list = [weight.clone() for weight in layer.w2_weight.data.unbind(dim=0)]
-                del layer.w13_weight
-                del layer.w2_weight
-                torch.npu.empty_cache()
+                if use_megamoe and is_950():
```
Comment (kunpengW-code) [inline vllm_ascend/ops/fused_moe/routed_experts.py]: Avoid using `is_950`. Please implement this via `hardware_profile`.

File: vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py [hunk 13]
```diff
@@ -302,12 +325,48 @@ def process_weights_after_loading(self, layer):
...
-        layer.w13_weight.data = layer.w13_weight.data.transpose(1, 2)
-        layer.w2_weight.data = layer.w2_weight.data.transpose(1, 2)
-        layer.w13_weight_scale.data = layer.w13_weight_scale.data.transpose(1, 2)
-        layer.w2_weight_scale.data = layer.w2_weight_scale.data.transpose(1, 2)
+        # MegaMoe (FUSED_MC2 on A5) expects the weights in their original
+        # (out, in) layout — weight1=(E, 2*inter, hidden//2), weight2=(E,
+        # hidden, inter//2) for packed FP4 — and the reshaped (not transposed)
+        # E8M0 scales. Capture per-expert lists before the non-mega path
+        # transposes everything below.
+        if use_cann_megamoe(get_current_vllm_config()) and is_950():
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py]: MXFP quantization already has built-in guards to block non-950 devices, so there is no need for an extra `is_950` check.

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py [hunk 14]
```diff
@@ -185,19 +188,91 @@ def apply(
...
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py]: ditto

File: vllm_ascend/ascend_config.py [hunk 15]
```diff
@@ -889,6 +890,10 @@ def _check_mix_placement(self):
 
     @staticmethod
     def _is_megamoe_supported_by_config(vllm_config: VllmConfig) -> bool:
+        if is_950():
```
Comment (kunpengW-code) [inline vllm_ascend/ascend_config.py]: The conditions for enabling MegaMoE on 950 differ significantly from A3. It is recommended to handle this distinction directly in `hardware_profile` by adding a separate config like `HardwareCapability.CANN_MEGAMOE_MXFP`.

File: vllm_ascend/ascend_forward_context.py (hunk 3 above)
Comment (hanxi-java) [reply vllm_ascend/ascend_forward_context.py:372]: Confirmed with the committer, the A5 MegaMoe operator does not support non‑quantized scenarios for the time being. The draft model judgment logic needs to be retained.

File: vllm_ascend/ascend_forward_context.py (hunk 10 above)
Comment (hanxi-java) [reply vllm_ascend/ascend_forward_context.py]: removed

File: vllm_ascend/ops/fused_moe/moe_comm_method.py (hunk 11 above)
Comment (hanxi-java) [reply vllm_ascend/ops/fused_moe/moe_comm_method.py]: modified

File: vllm_ascend/ops/fused_moe/routed_experts.py (hunk 12 above)
Comment (hanxi-java) [reply vllm_ascend/ops/fused_moe/routed_experts.py]: modified

File: vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py (hunk 13 above)
Comment (hanxi-java) [reply vllm_ascend/quantization/methods/w4a4/w4a4_mxfp4.py]: removed

File: vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py (hunk 14 above)
Comment (hanxi-java) [reply vllm_ascend/quantization/methods/w4a8/w4a8_mxfp4.py]: removed

File: vllm_ascend/ascend_config.py (hunk 15 above)
Comment (hanxi-java) [reply vllm_ascend/ascend_config.py]: modified

**Comment by hanxi-java** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Review by kunpengW-code — APPROVED** [review]
